qutip/core/cy/qobjevo.pyx

```diff
@@ -0,0 +1,839 @@
+            `trace(self @ matrix)` is `state` is an operator or operator-ket.
+        """
+        # TODO: remove reading from `settings` for a typed value when options
+        # support property.
+        cdef float herm_rtol = settings.core['rtol']
+        if not isinstance(state, Qobj):
+            raise TypeError("A Qobj state is expected")
+        if not (self.isoper or self.issuper):
+            raise ValueError("Must be an operator or super operator to compute"
+                             " an expectation value")
+        if not (
+            (self.dims[1] == state.dims[0]) or
+            (self.issuper and self.dims[1] == state.dims)
+        ):
+            raise ValueError("incompatible dimensions " + str(self.dims) +
+                             ", " + str(state.dims))
+        out = self.expect_data(t, state.data)
+        if out == 0 or (out.real and fabs(out.imag / out.real) < herm_rtol):
+            return out.real
+        return out
+
+    cpdef double complex expect_data(QobjEvo self, double t, Data state):
+        """
+        Expectation is defined as `state.adjoint() @ self @ state` if
+        `state` is a vector, or `state` is an operator and `self` is a
+        superoperator.  If `state` is an operator and `self` is an operator,
+        then expectation is `trace(self @ matrix)`.
+        """
+        if type(state) is Dense:
+            return self._expect_dense(t, state)
+        cdef _BaseElement part
+        cdef double complex out = 0., coeff
+        cdef Data part_data
+        cdef object expect_func
+        t = self._prepare(t, state)
+        if self.issuper:
+            if state.shape[1] != 1:
+                state = _data.column_stack(state)
+            expect_func = _data.expect_super
+        else:
+            expect_func = _data.expect
+
+        for element in self.elements:
+            part = (<_BaseElement> element)
+            coeff = part.coeff(t)
+            part_data = part.data(t)
+            out += coeff * expect_func(part_data, state)
+        return out
+
+    cdef double complex _expect_dense(QobjEvo self, double t, Dense state):
+        """For Dense state, `column_stack_dense` can be done inplace if in
+        fortran format."""
+        cdef size_t nrow = state.shape[0]
+        cdef _BaseElement part
+        cdef double complex out = 0., coeff
+        cdef Data part_data
+        t = self._prepare(t, state)
+        if self.issuper:
+            if state.shape[1] != 1:
+                state = column_stack_dense(state, inplace=state.fortran)
+            for element in self.elements:
+                part = (<_BaseElement> element)
+                coeff = part.coeff(t)
+                part_data = part.data(t)
+                out += coeff * expect_super_data_dense(part_data, state)
+            if state.fortran:
+                # `state` was reshaped inplace, restore it's original shape
+                column_unstack_dense(state, nrow, inplace=state.fortran)
```

Comment:
Am I understanding that we mutate `state` (if `state.fortran`) and then mutate it back here? What happens if one of the functions calls raises an exception? Will state be left in a modified state? Should we fix that somehow?

Reply:
We re not left with a modified state because of `cython`'s error handling. error are ignored and only caught back when in python:

```
>>> qevo = qt.QobjEvo(qt.spre(qt.qeye(4)))
>>> super_state = qt.rand_dm(3, dtype='dense').data.reorder()
>>> qevo.expect_data(0, super_state)

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~/algo/qutip/qutip/core/data/expect.pyx in qutip.core.data.expect.expect_super_csr_dense()

~/algo/qutip/qutip/core/data/expect.pyx in qutip.core.data.expect._check_shape_super()

ValueError: incompatible shapes (16, 16), (9, 1)

Exception ignored in: 'qutip.core.data.expect.expect_super_data_dense'
Traceback (most recent call last):
  File "qutip/core/data/expect.pyx", line 209, in qutip.core.data.expect.expect_super_csr_dense
  File "qutip/core/data/expect.pyx", line 37, in qutip.core.data.expect._check_shape_super
ValueError: incompatible shapes (16, 16), (9, 1)

0j

>>> np.isfortran(super_state.as_ndarray()), super_state.shape
(True, (3, 3))
```

As you can see the shape is still `(3,3)` while being dense fortran data.
(And `qevo.expect_data` returned `0j` with the error.)

Reply:
Woot. This Cython error handling stuff is crazy. :)

Reply:
That's not safe error handling! Ignoring the exception doesn't mean we should continue - when we're running in C mode, we have a responsibility to ensure that _no_ exceptions are ever raised. This is how segfaults happen.

Reply:
Just to add: the fact that you could make an "Exception ignored in" message appear using only Python-space commands means that's something's wrong.  Given that the internal function calls already acquire the GIL to raise exceptions, there shouldn't be too much worry about possibly needing to do that here as well.  The cost of checking the exception state after the function should be trivial compared to the cost of calculating even a single expectation value, and even if it isn't, Python-accessible segfaults aren't a worthwhile tradeoff for that extra bit of speed.

I think you'll have to check the error condition after each one in the loop.  Things like `column_stack_dense` can mutate the shape of a `Dense` in-place, so `QobjEvo` can't be certain its shape is always accurate.  Those mutations at least won't cause segfaults, but they're still Python-accessible errors that cannot be avoided, so we have to check and raise.

Reply:
Previously `expect_dense` had `except *` so perhaps we can fix this by adding that back and then by putting the clean up in a finally block?

Perhaps we can also add a test for a coefficient call that raises (which was the case I was originally worried about).